subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java

```diff
@@ -0,0 +1,150 @@
+import org.gradle.internal.progress.OperationFinishEvent;
+import org.gradle.internal.progress.OperationStartEvent;
+import org.gradle.internal.time.Clock;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Emitting build operations for tests
+ */
+public class TestListenerBuildOperationAdapter implements TestListenerInternal {
+    public static final TestOutputBuildOperationType.Details TEST_OUTPUT_DETAILS = new TestOutputBuildOperationType.Details() {
+    };
+    private final Map<TestDescriptor, InProgressExecuteTestBuildOperation> runningTests = new HashMap<TestDescriptor, InProgressExecuteTestBuildOperation>();
+    private final BuildOperationListener listener;
+    private final BuildOperationIdFactory buildOperationIdFactory;
+    private final Clock clock;
+    private final BuildOperationExecutor buildOperationExecutor;
+
+    public TestListenerBuildOperationAdapter(BuildOperationExecutor buildOperationExecutor, BuildOperationListener listener, BuildOperationIdFactory buildOperationIdFactory, Clock clock) {
+        this.buildOperationExecutor = buildOperationExecutor;
+        this.listener = listener;
+        this.buildOperationIdFactory = buildOperationIdFactory;
+        this.clock = clock;
+    }
+
+    @Override
+    public void started(final TestDescriptorInternal testDescriptor, TestStartEvent startEvent) {
+        BuildOperationDescriptor testBuildOperationDescriptor = createTestBuildOperationDescriptor(testDescriptor, startEvent);
+        long currentTime = clock.getCurrentTime();
+        runningTests.put(testDescriptor, new InProgressExecuteTestBuildOperation(testBuildOperationDescriptor, currentTime));
+        listener.started(testBuildOperationDescriptor, new OperationStartEvent(currentTime));
+    }
+
+    @Override
+    public void completed(TestDescriptorInternal testDescriptor, TestResult testResult, TestCompleteEvent completeEvent) {
+        InProgressExecuteTestBuildOperation runningOp = runningTests.remove(testDescriptor);
+        listener.finished(runningOp.descriptor, new OperationFinishEvent(runningOp.startTime, clock.getCurrentTime(), testResult.getException(), new BuildOperationTestResult(testResult)));
+    }
+
+    @Override
+    public void output(final TestDescriptorInternal testDescriptor, final TestOutputEvent event) {
+        InProgressExecuteTestBuildOperation runningOp = runningTests.get(testDescriptor);
+        BuildOperationDescriptor.Builder outputDescription = BuildOperationDescriptor.displayName(event.getDestination().name())
+            .operationType(BuildOperationCategory.TASK)
+            .details(TEST_OUTPUT_DETAILS);
+        BuildOperationDescriptor outputBuildOperationDescriptor = outputDescription.build(newOperationIdentifier(), runningOp.descriptor.getId());
+        long currentTime = clock.getCurrentTime();
+        listener.started(outputBuildOperationDescriptor, new OperationStartEvent(currentTime));
```

Comment:
❌ This doesn't really make sense to me - Starting and finishing a build operation for each piece of output a test creates. That would clutter the trace output pretty quickly. Wouldn't it be better to aggregate the output in the test finished event?

Reply:
We are currently missing the concept of a single point in time event in build operation land, like the progress events. When we move to having the plugin observe logging output via operations, we'll have the same problem.

Forcing Gradle to always buffer all of the test output in memory until the test finishes seems like a big sacrifice (fwiw, the scan plugin is already doing this but its a different thing doing it there in the plugin).

Alternatively, we enrich the operation event and notification APIs to support point-in-time events like the existing progress model. I'm leaning towards that.